pkg/autoscaler/autoscaler/tikv_autoscaler.go

```diff
@@ -24,38 +24,72 @@ import (
 	if tac.Spec.TiKV == nil {
+		emptyConsecutiveCount(tc, v1alpha1.TiKVMemberType)
 		return nil
 	}
 	sts, err := am.stsLister.StatefulSets(tc.Namespace).Get(operatorUtils.GetStatefulSetName(tc, v1alpha1.TiKVMemberType))
 	if err != nil {
 		return err
 	}
 	if !checkAutoScalingPrerequisites(tc, sts, v1alpha1.TiKVMemberType) {
+		emptyConsecutiveCount(tc, v1alpha1.TiKVMemberType)
 		return nil
 	}
-	targetReplicas := tc.Spec.TiKV.Replicas
-
-	// TODO: sync tikv .metrics from prometheus
-	// sum(rate(tikv_grpc_msg_duration_seconds_count{cluster="tidb", type!="kv_gc"}[1m])) by (instance)
-	//for _, _ = range tac.Spec.TiKV.Metrics {
-	//	// revive:disable:empty-block
-	//}
+	currentReplicas := getStateUpReplicas(tc)
+	targetReplicas := calculateRecommendedReplicas(tac, v1alpha1.TiKVMemberType, client)
 	targetReplicas = limitTargetReplicas(targetReplicas, tac, v1alpha1.TiKVMemberType)
 	if targetReplicas == tc.Spec.TiKV.Replicas {
+		emptyConsecutiveCount(tc, v1alpha1.TiKVMemberType)
+		return nil
+	}
+	return syncTiKVAfterCalculated(tc, tac, currentReplicas, targetReplicas, tc.Spec.TiKV.Replicas-currentReplicas)
+}
+
+// syncTiDBAfterCalculated would check the Consecutive count to avoid jitter, and it would also check the interval
+// duration between each auto-scaling. If either of them is not meet, the auto-scaling would be rejected.
+// If the auto-scaling is permitted, the timestamp would be recorded and the Consecutive count would be zeroed.
+// The currentReplicas of TiKV calculated in auto-scaling is the count of the StateUp TiKV instance, so we need to
+// add the number of other state tikv instance replicas when we update the TidbCluster.Spec.TiKV.Replicas
+func syncTiKVAfterCalculated(tc *v1alpha1.TidbCluster, tac *v1alpha1.TidbClusterAutoScaler, currentReplicas, recommendedReplicas, failureStoreCount int32) error {
+	if err := updateConsecutiveCount(tc, tac, v1alpha1.TiKVMemberType, currentReplicas, recommendedReplicas); err != nil {
+		return err
+	}
+
+	ableToScale, err := checkConsecutiveCount(tc, tac, v1alpha1.TiKVMemberType, currentReplicas, recommendedReplicas)
+	if err != nil {
+		return err
+	}
+	if !ableToScale {
 		return nil
 	}
+
 	intervalSeconds := tac.Spec.TiKV.ScaleInIntervalSeconds
-	if targetReplicas > tc.Spec.TiKV.Replicas {
+	if recommendedReplicas > tc.Spec.TiKV.Replicas {
 		intervalSeconds = tac.Spec.TiKV.ScaleOutIntervalSeconds
 	}
-	ableToScale, err := checkStsAutoScalingInterval(tc, *intervalSeconds, v1alpha1.TiKVMemberType)
+	ableToScale, err = checkStsAutoScalingInterval(tc, *intervalSeconds, v1alpha1.TiKVMemberType)
 	if err != nil {
 		return err
 	}
 	if !ableToScale {
 		return nil
 	}
-	tc.Spec.Annotations[label.AnnTiKVLastAutoScalingTimestamp] = time.Now().String()
-	tc.Spec.TiKV.Replicas = targetReplicas
+	updateTcTiKVAnnIfScale(tc)
+	tc.Spec.TiKV.Replicas = recommendedReplicas + failureStoreCount
```

Comment:
failover function will add failure stores again.

Reply:
The recommended replicas is calculated by `currentReplicas` which is excluded the failure replicas. That's why we should add the failure replicas into recommended replicas.

Reply:
updated.